### DISCUSSION TITLE
Skip top-level "items" in graphql

### DISCUSSION THREAD

**Discussion by emregency** [directus/directus#4265] [Feature Requests]
In the GraphQL realm, the "folders" without attributes don't exist. Everything is normally content in graphs. The `items/<collection-name>` is more an anti-pattern in graph-thinking as `items` doesn't exist as a `node`; it has no attributes.

**Suggestion 1**
To keep the consistency with the current [`system` collections](https://docs.directus.io/reference/api/graphql/#querying-data) that are exposed directly, does it make sense to get rid of the `items` level and query directly the user-defined collections?

```
query {
  <collection-name>
     {
        id 
        <attribute-1>
     }
}
```
**Suggestion 2**
Add attributes to `items` e.g. id, description...

**Comment by rijkvanzanten** [comment]
> does it make sense to get rid of the items level and query directly the user-defined collections?

I'd rather not, it'll mean that we either have to start prefixing the system stuff with `directus_` to match the collection name (which goes against our white-labeling principles), or it'd be a massive source of naming conflicts.

> In the GraphQL realm, the "folders" without attributes don't exist. Everything is normally content in graphs. The items/<collection-name> is more an anti-pattern in graph-thinking as items doesn't exist as a node; it has no attributes.

What do you mean by _the "folders" without attributes don't exist_ and _Everything is normally content_? Do you have some examples / reading to clarify this?

I don't have a ton of experience in GraphQL, but from what I can tell most other GraphQL endpoints in platforms like these just... seem to do whatever works?

**Reply by emregency** [reply]
Sorry for the confusion.

What I mean is:

A user-defined collection let's say `products` "under" `items` is actually in GraphQL implementing an interface of `items` just like typescript. But what are the properties of `items`...

For example:

```graphql
interface Items {
    id: ID!
    created_at: Date!
    updated_by: User
}

type Products implements Items {
    id: ID!
    created_at: Date!
    updated_by: User
    // ... all the fields that are created by the user
}
```
**And Not**

```graphql
type items {
    products: [Products!]
}
```

The first example is from Strapi. The collection `Restaurants` is a user-created collection, its REST API is ([here](https://strapi.io/documentation/developer-docs/latest/content-api/api-endpoints.html#api-endpoints)). And it is also a `type` in GraphQL as described [here](https://strapi.io/documentation/developer-docs/latest/getting-started/graphql.html#fetch-your-restaurant-collection-type).

GraphCMS works the same way as Strapi. An example of a model creation is [here](https://graphcms.com/docs/schema/models)

DGraph "native GraphQL database" has 2 endpoints: one for user-created content and another for admin [link here](https://dgraph.io/docs/graphql/admin/#endpoints)

All types are objects that have attributes. They can all be queried (sometimes single, sometimes plural), mutated, subscribed to, filtered, sorted, etc... 

I have two questions: 
1- Why do we treat `Files` or `Users` differently from `Products` or `Restaurants`?
2- Can we separate system and items as two endpoints like DGraph did?

**Reply by rijkvanzanten** [reply]
> 1- Why do we treat Files or Users differently from Products or Restaurants?

Because the former has a bunch of custom functionality / logic associated that doesn't apply to the latter. Everything under "items" has a consistent API, while functionality will change between the different system collections.

> 2- Can we separate system and items as two endpoints like DGraph did?

We could, but I'd rather not. What we have now still feels cleaner to me 🤷🏻

**Reply by rijkvanzanten** [reply]
I still don't really understand what you're saying 😞 

> A user-defined collection let's say products "under" items is actually in GraphQL implementing an interface of items just like typescript. But what are the properties of items...

The problem is that I don't really understand what's inherently wrong with the

```graphql
type items {
    products: [Products!]
}
```

approach. To me it very clearly scopes functionality and expected API within a root group. 🤷🏻 

(Similarly, I personally find the mutation style where an action is prepended to a model name (eg `posts` -> `updatePost` abhorrent). I'd personally find something like "in posts, do update" cleaner)

**Reply by emregency** [reply]
> The problem is that I don't really understand what's inherently wrong with the
> ```
>type items {
>   products: [Products!]
>}
>```
> approach. To me it very clearly scopes functionality and expected API within a root group. 🤷🏻

This is **exactly** achieved through [interfaces ](https://graphql.org/learn/schema/#interfaces) and not through attributes. The notation above doesn't scope nor says the API that is expected, rather says "I have items that has an attribute called products that link to Products" and does not say "I have Products that are of type Items". 

The official documentation's example:

> An Interface is an abstract type that includes a certain set of fields that a type must include to implement the interface.
> 
> For example, you could have an interface Character that represents any character in the Star Wars trilogy:
>
> ```
> interface Character {
>   id: ID!
>   name: String!
>   friends: [Character]
>   appearsIn: [Episode]!
> }
> ```
> This means that any type that implements Character needs to have these exact fields, with these arguments and return types.
> 
> For example, here are some types that might implement Character:
> 
> ```
> type Human implements Character {
>   id: ID!
>   name: String!
>   friends: [Character]
>   appearsIn: [Episode]!
>   starships: [Starship]
>   totalCredits: Int
> }
> ```
>  
> ```
> type Droid implements Character {
>   id: ID!
>   name: String!
>   friends: [Character]
>   appearsIn: [Episode]!
>   primaryFunction: String
> }
> ```
> You can see that both of these types have all of the fields from the Character interface, but also bring in extra fields, totalCredits, starships and primaryFunction, that are specific to that particular type of character.

**Reply by WoLfulus** [reply]
> The problem is that I don't really understand what's inherently wrong with the

This actually makes sense to me, what it's what I'd expect too.

Also, this is one of the reasons why I separated the data (`/graphql`) from the system (`/graphql/system`) in the Laravel project. This really annoyed me at the time, and it's the same situation we discussed this week regarding what people expect vs what we want to/can do.

Imho we could take some time to discuss things like this since we already have some pending iterations over the graphql implementation.

Let me know!

**Reply by rijkvanzanten** [reply]
> This actually makes sense to me, what it's what I'd expect too.

All I'm looking for is an authoritative source stating _why_ it's the expected thing 😄  

Like I mentioned before, my limit experience in GraphQL isn't enough to call the shots, but I can't commit to any definitive changes based on individual feelings alone either. If there's some sort of spec we can follow (which we can write ourselves too), then I'll gladly concede

**Reply by emregency** [reply]
@WoLfulus 

> This really annoyed me at the time,

I am not a big fan of 2 endpoints neither. 

> This actually makes sense to me, what it's what I'd expect too.

I must ask you guys: **What is the primary key/id of the `items` object then?**


> To provide options for GraphQL clients to elegantly handle caching and data refetching, GraphQL servers need to expose object identifiers in a standardized way.
> For this to work, a client will need to query via a standard mechanism to request an object by ID. Then, in the response, the schema will need to provide a standard way of providing these IDs. 

source: https://graphql.org/learn/global-object-identification/

Also referred in GitLab's GraphQL API style guide https://docs.gitlab.com/ee/development/api_graphql_styleguide.html#global-ids

**Reply by rijkvanzanten** [reply]
> I must ask you guys: What is the primary key/id of the `items` object then?

There is none, but I don't see how that's different from fetching things relationally through another node? (eg `query { articles { authors { id }}}`

Also, keep in mind that as long as we allow auto-increment/string based keys, there won't be a globally unique identifier in Directus

**Reply by emregency** [reply]
> (eg query { articles { authors { id }}}

The query above is not a "valid" query in GraphQL. The valid version of it is:

```
query {
    articles {
        id
        authors {
            id
        }
     }
}
```

It is different because `articles` are objects that must have also an id. The user can also query all the `articles`:

```
query {
    articles {
        id
        title
     }
}
```

or filter, sort, limit.. articles:

```
query {
    articles(filter:  "", sort: "", limit: "") {
        id
        title
     }
}
```

probably `articles` implements something like that:

```
interface Items {
    id: ID!
    title: string!
....
}

type Articles implements Items {
   id: ID!
   title: string!
   authors: [Author!]
}
```

`items` cannot do it but `files`, `users` can. 

> Also, keep in mind that as long as we allow auto-increment/string based keys, there won't be a globally unique identifier in Directus

As I use UUID, I think I wouldn't have this problem. We can add this in the docs or during the ID type selection during collection creation.

**Reply by rijkvanzanten** [reply]
> The query above is not a "valid" query in GraphQL

Do you have a source for that? GraphQL seems to accept it perfectly fine when I query things like that 🤷🏻

**Reply by emregency** [reply]
It works indeed. That’s why I quoted the “valid”. An identifier, as far as I saw and tried to refer in the links, is implemented as a good practice. Especially for clients that manage a cache.

**Comment by emregency** [comment]
> (Similarly, I personally find the mutation style where an action is prepended to a model name (eg posts -> updatePost abhorrent). I'd personally find something like "in posts, do update" cleaner)

The mutations such as `updatePost` is a "convention" that is implemented as far as I have seen everywhere. The [the official docs](https://graphql.org/learn/queries/#mutations) mentions the same approach.

**Reply by emregency** [reply]
I leave here 3 naming conventions for mutations. Not authoritative, but still interesting.

https://www.graphile.org/postgraphile/crud-mutations/
https://hasura.io/docs/latest/graphql/core/databases/postgres/mutations/index.html
https://docs.gitlab.com/ee/development/api_graphql_styleguide.html#naming-conventions

**Comment by emregency** [comment]
"Data Normalization" in the cache is enabled through a unique identifier. And actually, I am running into the same issue, that's why I opened this discussion initially. When I query different collections, my cache is reset **every time** as `items` doesn't have a unique identifier.

🥇 Relay requires an `id` https://relay.dev/docs/en/thinking-in-graphql#caching-a-graph
🥈 Apollo Client requires an `id` https://www.apollographql.com/docs/android/essentials/normalized-cache/#data-normalization
🥉 urql requires an `id` https://formidable.com/open-source/urql/docs/graphcache/normalized-caching/#storing-normalized-data

**Reply by rijkvanzanten** [reply]
As per that relay example, it doesn't seem like globally unique IDs are required (i'm assuming it maps type+id instead of just id).

Am I right in understanding that to "solve" for these caching layers, all we'd need to add would be a `id` attribute that selects the single record?

**Reply by emregency** [reply]
Indeed, Relay doesn’t require a UUID. __typename+id works for the others too.

An ‘id’ would already be excellent for this cache issue.

**Reply by rijkvanzanten** [reply]
> When I query different collections, my cache is reset every time as items doesn't have a unique identifier.

Could you provide a repo with a minimal setup to test this with?

**Reply by emregency** [reply]
🤔 let me see what I can share.

**Reply by rijkvanzanten** [reply]
@emregency Any updates? I'd love to finalize this new spec before calling v9.0.0 complete, otherwise it'll be a breaking change 😄

**Reply by eikaramba** [reply]
i hope i understand it correctly that is the reason why i cannot do
```
query {
   article (id:"1") {
      id
   }
}
```
and just get the article with id 1 as an object back. currently i always need to use filters and then get an array with size 1 back. hopefully i found the correct discussion. would be nice to have single entity support as well

**Reply by rijkvanzanten** [reply]
@eikaramba use `article_by_id(id: ID)` (https://docs.directus.io/reference/items/#get-item-by-id)

**Comment by emregency** [comment]
@WoLfulus @rijkvanzanten would something like [KeystoneJS](https://www.keystonejs.com) be interesting for us (for mutations I thought)?

**Reply by rijkvanzanten** [reply]
Keystone is a whole different CMS, so using that as part of Directus might be a little rough. That being said, we can learn a lot from how they're handling the mutation generation based on the input type

**Comment by emregency** [comment]
Hey @rijkvanzanten . I started a new opportunity this week and it was a busy one. Sorry for the radio silence. I checked it and we get the warning of not having an identifier with Apollo client 3 and a react front end. 

But we dropped graphql in the meantime as we couldn’t do a “deep” filter. We are using the SDK since then. 

That’s why, it is overtaken by events for us.

**Reply by rijkvanzanten** [reply]
No worries!

As for 

> But we dropped graphql in the meantime as we couldn’t do a “deep” filter

This was actually just implemented and released (10 minutes ago)! 😄

**Comment by iksent** [comment]
I am experiencing the same issue, here is the details:
https://stackoverflow.com/questions/66617742/how-to-configure-apollo-cache-for-specific-graphql-queries

Apollo Cache isn't working by default for "items" queries, because items have no IDs inside it and I can't understand how to configure it.

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
